### PR TITLE
Implement base protocol class

### PR DESCRIPTION
**Pull request by asvetlov**
To avoid a dependency from internal `asyncio.streams.FlowControlMixin`

### CODE REVIEW

**Comment by kxepal** [comment]
> To avoid a dependency from internal asyncio.streams.FlowControlMixin

Why to avoid this? Do you expect that internals will change in some breaking way soon?

**Comment by asvetlov** [comment]
The class is not a part of public API. Technically it should be prefixed by underscore but we've missed it.

The second reason is type hinting: we have a public `.transport` property and expose our protocols in our public API (not sure if we should do it but the change can be done in 4.0 only).
Marking a property as "returning `asyncio.Protocol`" crashes mypy.

**Comment by kxepal** [comment]
Got it. Thanks!

File: aiohttp/base_protocol.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+class BaseProtocol(asyncio.Protocol):
+    def __init__(self, loop=None):
+        if loop is None:
+            self._loop = asyncio.get_event_loop()
+        else:
+            self._loop = loop
+        self._paused = False
+        self._drain_waiter = None
+        self._connection_lost = False
+        self.transport = None
```
Comment (kxepal) [inline aiohttp/base_protocol.py:15]: Do we want transport as public attribute or instead public read-only property? I think transport is only valid to set via `connection_made` method, not somehow adhoc'y.

Comment (asvetlov) [reply aiohttp/base_protocol.py:15]: I'm trying don't touch transport attr, many aiohttp codes use it internally.
Tests modify `.transport` wildly as well.
Better to deprecate `.protocol` at all in our public API.

Comment (kxepal) [reply aiohttp/base_protocol.py:15]: :+1:

**Pull request merged**